sdk/eventhub/Azure.Messaging.EventHubs/tests/Snippets/Sample10_MockingReadingEventsTests.cs

```diff
@@ -0,0 +1,292 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs.Consumer;
+using Azure.Messaging.EventHubs.Primitives;
+using Azure.Messaging.EventHubs.Producer;
+using Moq;
+using NUnit.Framework;
+
+namespace Azure.Messaging.EventHubs.Tests.Snippets
+{
+    [TestFixture]
+    public class Sample10_MockingReadingEventsTests
+    {
+        /// <summary>
+        ///   Performs basic unit test validation of the contained snippet.
+        /// </summary>
+        ///
+        [Test]
+        public async Task MockingReadEventsFromPartition()
+        {
+            #region Snippet:EventHubs_Sample10_MockingReadEventsFromPartition
+
+            var mockEventHubConsumerClient = new Mock<EventHubConsumerClient>();
+            var mockPartitionId = "sample partition id";
+
+            mockEventHubConsumerClient
+                .Setup(c => c.ReadEventsFromPartitionAsync(It.IsAny<string>(), It.IsAny<EventPosition>(), It.IsAny<CancellationToken>()))
+                .Returns(GetMockEventsFromPartition(mockPartitionId));
+
+            var consumer = mockEventHubConsumerClient.Object;
+            try
+            {
+                using CancellationTokenSource cancellationSource = new CancellationTokenSource();
+                cancellationSource.CancelAfter(TimeSpan.FromSeconds(30));
+
+                EventPosition startingPosition = EventPosition.Earliest;
+
+                await foreach (PartitionEvent partitionEvent in consumer.ReadEventsFromPartitionAsync(
+                    mockPartitionId,
+                    startingPosition,
+                    cancellationSource.Token))
+                {
+                    string readFromPartition = partitionEvent.Partition.PartitionId;
+                    ReadOnlyMemory<byte> eventBodyBytes = partitionEvent.Data.EventBody.ToMemory();
+                }
+            }
+            catch (TaskCanceledException)
+            {
+                // This is expected if the cancellation token is
+                // signaled.
+            }
+            finally
+            {
+                await consumer.CloseAsync();
+            }
+
+            #endregion
+        }
+
+        /// <summary>
+        ///   Performs basic unit test validation of the contained snippet.
+        /// </summary>
+        ///
+        [Test]
+        public async Task MockingReadEventsFromPartition_WithLastEnqueuedEventProperties()
+        {
```

Comment:
added a new test method to demonstrate reading events with last enqueued event properties, this uses the `GetsMockEventFromPartition` overload with `lastEnqueuedEventPartitions`

Reply:
let me know if we need this test method or just the local `GetsMockEventFromPartition` overload with `lastEnqueuedEventPartitions` is enough.

Reply:
If we make a local function out of `GetMockEventsFromPartition` and demonstrate calling it with the last enqueued mock, I think that's a good way to focus on what we're trying to highlight here.   We can probably kill the consumer bits, call `GetMockEventsFromPartition`  and demonstrate using the partition context to return the mocked data.   Maybe we consider a trailing comment on being able to use this pattern while mocking out consumers.

What are your thoughts?

Reply:
just reiterating/ thinking out loud here : 
we should not mock the consumer, and call the `GetMockEventsFromPartition()` directly and use that to return the mocked data using the returned partition context?

What scenario's would one want to not use this pattern?
Is the idea to demonstrate using `PartitionContext` to fetch mock data w/o using the `ConsumerClient`?

Reply:
We want each snippet to be as self-contained as possible; ideally, a developer should be able to copy/paste the sample and it should compile/run without having to hunt down common utility methods in other snippets.  For clarity, it's better to demonstrate a concept in isolation than to introduce complexity with trying to demonstrate too many things at once.

If we feel that showing the generation doesn't work without the consumer bits, I'm good with that.   We'll want to break things down into multiple snippets that we can render back-toback since there's a clear functionality delineation, which also potentially allows us to reuse the same utility method snippet elsewhere.  For example:

```csharp
#region Snippet:EventHubs_Sample10_GeneratingMockEvents
public async IAsyncEnumerable<PartitionEvent> GetMockEventsFromPartition(int partitionId)
{   
    var partitionContext = EventHubsModelFactory.PartitionContext(partitionId,);

    await Task.Yield();
    yield return new PartitionEvent(partitionContext, new EventData("One"));
    yield return new PartitionEvent(partitionContext, new EventData("Two"));
    yield return new PartitionEvent(partitionContext, new EventData("Three"));
}
#endregion
```
```csharp
#region Snippet:EventHubs_Sample10_UsingMockEventsWithTheConsumer
// .. THE CONSUMER BITS
#endregion
```

For a more complex illustration, we break things down in a couple of different places, in different ways:

- In our basic [Hello World](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample01_HelloWorld.md) sample, we introduce creating clients in their own snippet and mention that we're assuming this pattern below for the more complex cases.   This works because it's simple and foundational.

- In the [Migration Guide](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs/MigrationGuide.md#migrating-eventprocessorhost-checkpoints), we demonstrate moving checkpoints in a few snippets that appear together in a section, where each builds on the other.   We do this to try and reduce complexity by talking over independent parts independently.  Here, things are tightly clustered because they are somewhat complex and their usage is really isolated to this particular scenario.